docs/usage/annotations.rst

```diff
@@ -144,3 +201,84 @@ To do so, one can declare an alias to the reusable qubits.
 .. note::
     Even if the reusable alias is not used in code, an alias must be defined to mark qubits as reusable.
     The identifier is insignificant.
+
+Entangled & Dirty Ancillae
+--------------------------
+If qubits are used within a snippet and are not annotated with either ``@leqo.output`` or ``@leqo.reusable``,
+they are treated by that snippet as entangled ancillae.
+Subsequent snippets interpret these qubits as dirty ancillae.
+Dirty ancillae may exist in an arbitrary quantum state, potentially entangled with other qubits,
+and must be explicitly annotated with ``@leqo.dirty`` to signify their intentional reuse in a snippet.
+
+To use dirty ancillae within a snippet, the programmer must explicitly opt in by annotating the qubit declaration with ``@leqo.dirty``.
+
+* The ``@leqo.dirty`` annotation follows the same implementation rules as input definitions, but omits indexing, as defined in :ref:`input definition <input-anker>`
+
+.. warning::
+    The state of a dirty ancilla qubit can be altered temporarily but must be restored at the end of a snippet.
+    Therefore measuring or uncompute a dirty qubit is not permitted.
+
+.. code-block:: openqasm3
+    :linenos:
+
+    // Single dirty ancilla
+    @leqo.dirty
+    qubit singleDirtyAncilla;
+
+.. code-block:: openqasm3
+    :linenos:
+
+    // Dirty ancilla array
+    @leqo.dirty
+    qubit[<<length>>] dirtyAncillaArray;
+
+Uncomputation
+-------------
+When modified clean ancillae or linking qubits can be uncomputed, the programmer may provide an explicit uncomputation block to reverse their effects.
+This is done using the ``@leqo.uncompute`` annotation, which defines a scoped region that is disabled by default via an ``if (false)`` statement.
+The compiler may override this statement to ``true`` if uncomputation of the associated qubits is determined to be necessary.
+
+A qubit annotated with ``@leqo.reusable`` within such a block is referred to as an uncomputable ancilla.
+
+* The ``@leqo.uncompute`` annotation must appear directly above an ``if (false)`` statement with a block body that must not be followed by an ``else`` statement
+* ``@leqo.uncompute`` annotations may appear multiple times in a program, each time referring to different uncomputation logic
+* Nested ``@leqo.uncompute`` if-blocks are not allowed
+* A ``@leqo.uncompute`` if-block must be declared at global scope
+* The ``@leqo.uncompute`` block must reverse all transformations on the associated qubit, removing entanglement and restoring each to the ``|0⟩`` state
+* ``@leqo.uncompute`` blocks only operate on existing variables, qubits or selfdeclared aliases
+* A ``@leqo.uncompute`` if-block must declare the uncomputed ancillae as reusable qubits by using the corresponding ``@leqo.reusable`` annotation
+
+.. warning::
+    Qubits previously annotated with ``@leqo.dirty`` must not be uncomputed
+
+.. note::
+    Not all operations are reversible; in such cases, the qubit should not be annotated as reusable.
```

Comment:
All quantum operations are per definition reversible (though an uncompute might still be hard). And, of course, measuremments are not reversible.